results/Log_2021-06-11_02-47-11/parameters.txt
# -----------------------------------#
# Parameters of the training session #
# -----------------------------------#

# Input parameters
# ****************

dataset = SemanticKitti
dataset_task = slam_segmentation
num_classes = 20
in_points_dim = 3
in_features_dim = 2
in_radius = 4.000000
input_threads = 10

# Model parameters
# ****************

architecture = simple resnetb resnetb_strided resnetb resnetb resnetb_strided resnetb_deformable resnetb_deformable resnetb_deformable_strided resnetb_deformable resnetb_deformable resnetb_deformable_strided resnetb_deformable resnetb_deformable nearest_upsample unary nearest_upsample unary nearest_upsample unary nearest_upsample unary
equivar_mode = 
invar_mode = 
num_layers = 5
first_features_dim = 128
use_batch_norm = 1
batch_norm_momentum = 0.020000

segmentation_ratio = 1.000000

# KPConv parameters
# *****************

first_subsampling_dl = 0.060000
num_kernel_points = 15
conv_radius = 2.500000
deform_radius = 6.000000
fixed_kernel_points = center
KP_extent = 1.200000
KP_influence = linear
aggregation_mode = sum
modulated = 0
n_frames = 1
max_in_points = 5537

max_val_points = 7895

val_radius = 4.000000

# Training parameters
# *******************

learning_rate = 0.010000
momentum = 0.980000
lr_decay_epochs = 1:0.984767 2:0.984767 3:0.984767 4:0.984767 5:0.984767 6:0.984767 7:0.984767 8:0.984767 9:0.984767 10:0.984767 11:0.984767 12:0.984767 13:0.984767 14:0.984767 15:0.984767 16:0.984767 17:0.984767 18:0.984767 19:0.984767 20:0.984767 21:0.984767 22:0.984767 23:0.984767 24:0.984767 25:0.984767 26:0.984767 27:0.984767 28:0.984767 29:0.984767 30:0.984767 31:0.984767 32:0.984767 33:0.984767 34:0.984767 35:0.984767 36:0.984767 37:0.984767 38:0.984767 39:0.984767 40:0.984767 41:0.984767 42:0.984767 43:0.984767 44:0.984767 45:0.984767 46:0.984767 47:0.984767 48:0.984767 49:0.984767 50:0.984767 51:0.984767 52:0.984767 53:0.984767 54:0.984767 55:0.984767 56:0.984767 57:0.984767 58:0.984767 59:0.984767 60:0.984767 61:0.984767 62:0.984767 63:0.984767 64:0.984767 65:0.984767 66:0.984767 67:0.984767 68:0.984767 69:0.984767 70:0.984767 71:0.984767 72:0.984767 73:0.984767 74:0.984767 75:0.984767 76:0.984767 77:0.984767 78:0.984767 79:0.984767 80:0.984767 81:0.984767 82:0.984767 83:0.984767 84:0.984767 85:0.984767 86:0.984767 87:0.984767 88:0.984767 89:0.984767 90:0.984767 91:0.984767 92:0.984767 93:0.984767 94:0.984767 95:0.984767 96:0.984767 97:0.984767 98:0.984767 99:0.984767 100:0.984767 101:0.984767 102:0.984767 103:0.984767 104:0.984767 105:0.984767 106:0.984767 107:0.984767 108:0.984767 109:0.984767 110:0.984767 111:0.984767 112:0.984767 113:0.984767 114:0.984767 115:0.984767 116:0.984767 117:0.984767 118:0.984767 119:0.984767 120:0.984767 121:0.984767 122:0.984767 123:0.984767 124:0.984767 125:0.984767 126:0.984767 127:0.984767 128:0.984767 129:0.984767 130:0.984767 131:0.984767 132:0.984767 133:0.984767 134:0.984767 135:0.984767 136:0.984767 137:0.984767 138:0.984767 139:0.984767 140:0.984767 141:0.984767 142:0.984767 143:0.984767 144:0.984767 145:0.984767 146:0.984767 147:0.984767 148:0.984767 149:0.984767 150:0.984767 151:0.984767 152:0.984767 153:0.984767 154:0.984767 155:0.984767 156:0.984767 157:0.984767 158:0.984767 159:0.984767 160:0.984767 161:0.984767 162:0.984767 163:0.984767 164:0.984767 165:0.984767 166:0.984767 167:0.984767 168:0.984767 169:0.984767 170:0.984767 171:0.984767 172:0.984767 173:0.984767 174:0.984767 175:0.984767 176:0.984767 177:0.984767 178:0.984767 179:0.984767 180:0.984767 181:0.984767 182:0.984767 183:0.984767 184:0.984767 185:0.984767 186:0.984767 187:0.984767 188:0.984767 189:0.984767 190:0.984767 191:0.984767 192:0.984767 193:0.984767 194:0.984767 195:0.984767 196:0.984767 197:0.984767 198:0.984767 199:0.984767 200:0.984767 201:0.984767 202:0.984767 203:0.984767 204:0.984767 205:0.984767 206:0.984767 207:0.984767 208:0.984767 209:0.984767 210:0.984767 211:0.984767 212:0.984767 213:0.984767 214:0.984767 215:0.984767 216:0.984767 217:0.984767 218:0.984767 219:0.984767 220:0.984767 221:0.984767 222:0.984767 223:0.984767 224:0.984767 225:0.984767 226:0.984767 227:0.984767 228:0.984767 229:0.984767 230:0.984767 231:0.984767 232:0.984767 233:0.984767 234:0.984767 235:0.984767 236:0.984767 237:0.984767 238:0.984767 239:0.984767 240:0.984767 241:0.984767 242:0.984767 243:0.984767 244:0.984767 245:0.984767 246:0.984767 247:0.984767 248:0.984767 249:0.984767 250:0.984767 251:0.984767 252:0.984767 253:0.984767 254:0.984767 255:0.984767 256:0.984767 257:0.984767 258:0.984767 259:0.984767 260:0.984767 261:0.984767 262:0.984767 263:0.984767 264:0.984767 265:0.984767 266:0.984767 267:0.984767 268:0.984767 269:0.984767 270:0.984767 271:0.984767 272:0.984767 273:0.984767 274:0.984767 275:0.984767 276:0.984767 277:0.984767 278:0.984767 279:0.984767 280:0.984767 281:0.984767 282:0.984767 283:0.984767 284:0.984767 285:0.984767 286:0.984767 287:0.984767 288:0.984767 289:0.984767 290:0.984767 291:0.984767 292:0.984767 293:0.984767 294:0.984767 295:0.984767 296:0.984767 297:0.984767 298:0.984767 299:0.984767 300:0.984767 301:0.984767 302:0.984767 303:0.984767 304:0.984767 305:0.984767 306:0.984767 307:0.984767 308:0.984767 309:0.984767 310:0.984767 311:0.984767 312:0.984767 313:0.984767 314:0.984767 315:0.984767 316:0.984767 317:0.984767 318:0.984767 319:0.984767 320:0.984767 321:0.984767 322:0.984767 323:0.984767 324:0.984767 325:0.984767 326:0.984767 327:0.984767 328:0.984767 329:0.984767 330:0.984767 331:0.984767 332:0.984767 333:0.984767 334:0.984767 335:0.984767 336:0.984767 337:0.984767 338:0.984767 339:0.984767 340:0.984767 341:0.984767 342:0.984767 343:0.984767 344:0.984767 345:0.984767 346:0.984767 347:0.984767 348:0.984767 349:0.984767 350:0.984767 351:0.984767 352:0.984767 353:0.984767 354:0.984767 355:0.984767 356:0.984767 357:0.984767 358:0.984767 359:0.984767 360:0.984767 361:0.984767 362:0.984767 363:0.984767 364:0.984767 365:0.984767 366:0.984767 367:0.984767 368:0.984767 369:0.984767 370:0.984767 371:0.984767 372:0.984767 373:0.984767 374:0.984767 375:0.984767 376:0.984767 377:0.984767 378:0.984767 379:0.984767 380:0.984767 381:0.984767 382:0.984767 383:0.984767 384:0.984767 385:0.984767 386:0.984767 387:0.984767 388:0.984767 389:0.984767 390:0.984767 391:0.984767 392:0.984767 393:0.984767 394:0.984767 395:0.984767 396:0.984767 397:0.984767 398:0.984767 399:0.984767 400:0.984767 401:0.984767 402:0.984767 403:0.984767 404:0.984767 405:0.984767 406:0.984767 407:0.984767 408:0.984767 409:0.984767 410:0.984767 411:0.984767 412:0.984767 413:0.984767 414:0.984767 415:0.984767 416:0.984767 417:0.984767 418:0.984767 419:0.984767 420:0.984767 421:0.984767 422:0.984767 423:0.984767 424:0.984767 425:0.984767 426:0.984767 427:0.984767 428:0.984767 429:0.984767 430:0.984767 431:0.984767 432:0.984767 433:0.984767 434:0.984767 435:0.984767 436:0.984767 437:0.984767 438:0.984767 439:0.984767 440:0.984767 441:0.984767 442:0.984767 443:0.984767 444:0.984767 445:0.984767 446:0.984767 447:0.984767 448:0.984767 449:0.984767 450:0.984767 451:0.984767 452:0.984767 453:0.984767 454:0.984767 455:0.984767 456:0.984767 457:0.984767 458:0.984767 459:0.984767 460:0.984767 461:0.984767 462:0.984767 463:0.984767 464:0.984767 465:0.984767 466:0.984767 467:0.984767 468:0.984767 469:0.984767 470:0.984767 471:0.984767 472:0.984767 473:0.984767 474:0.984767 475:0.984767 476:0.984767 477:0.984767 478:0.984767 479:0.984767 480:0.984767 481:0.984767 482:0.984767 483:0.984767 484:0.984767 485:0.984767 486:0.984767 487:0.984767 488:0.984767 489:0.984767 490:0.984767 491:0.984767 492:0.984767 493:0.984767 494:0.984767 495:0.984767 496:0.984767 497:0.984767 498:0.984767 499:0.984767 500:0.984767 501:0.984767 502:0.984767 503:0.984767 504:0.984767 505:0.984767 506:0.984767 507:0.984767 508:0.984767 509:0.984767 510:0.984767 511:0.984767 512:0.984767 513:0.984767 514:0.984767 515:0.984767 516:0.984767 517:0.984767 518:0.984767 519:0.984767 520:0.984767 521:0.984767 522:0.984767 523:0.984767 524:0.984767 525:0.984767 526:0.984767 527:0.984767 528:0.984767 529:0.984767 530:0.984767 531:0.984767 532:0.984767 533:0.984767 534:0.984767 535:0.984767 536:0.984767 537:0.984767 538:0.984767 539:0.984767 540:0.984767 541:0.984767 542:0.984767 543:0.984767 544:0.984767 545:0.984767 546:0.984767 547:0.984767 548:0.984767 549:0.984767 550:0.984767 551:0.984767 552:0.984767 553:0.984767 554:0.984767 555:0.984767 556:0.984767 557:0.984767 558:0.984767 559:0.984767 560:0.984767 561:0.984767 562:0.984767 563:0.984767 564:0.984767 565:0.984767 566:0.984767 567:0.984767 568:0.984767 569:0.984767 570:0.984767 571:0.984767 572:0.984767 573:0.984767 574:0.984767 575:0.984767 576:0.984767 577:0.984767 578:0.984767 579:0.984767 580:0.984767 581:0.984767 582:0.984767 583:0.984767 584:0.984767 585:0.984767 586:0.984767 587:0.984767 588:0.984767 589:0.984767 590:0.984767 591:0.984767 592:0.984767 593:0.984767 594:0.984767 595:0.984767 596:0.984767 597:0.984767 598:0.984767 599:0.984767 600:0.984767 601:0.984767 602:0.984767 603:0.984767 604:0.984767 605:0.984767 606:0.984767 607:0.984767 608:0.984767 609:0.984767 610:0.984767 611:0.984767 612:0.984767 613:0.984767 614:0.984767 615:0.984767 616:0.984767 617:0.984767 618:0.984767 619:0.984767 620:0.984767 621:0.984767 622:0.984767 623:0.984767 624:0.984767 625:0.984767 626:0.984767 627:0.984767 628:0.984767 629:0.984767 630:0.984767 631:0.984767 632:0.984767 633:0.984767 634:0.984767 635:0.984767 636:0.984767 637:0.984767 638:0.984767 639:0.984767 640:0.984767 641:0.984767 642:0.984767 643:0.984767 644:0.984767 645:0.984767 646:0.984767 647:0.984767 648:0.984767 649:0.984767 650:0.984767 651:0.984767 652:0.984767 653:0.984767 654:0.984767 655:0.984767 656:0.984767 657:0.984767 658:0.984767 659:0.984767 660:0.984767 661:0.984767 662:0.984767 663:0.984767 664:0.984767 665:0.984767 666:0.984767 667:0.984767 668:0.984767 669:0.984767 670:0.984767 671:0.984767 672:0.984767 673:0.984767 674:0.984767 675:0.984767 676:0.984767 677:0.984767 678:0.984767 679:0.984767 680:0.984767 681:0.984767 682:0.984767 683:0.984767 684:0.984767 685:0.984767 686:0.984767 687:0.984767 688:0.984767 689:0.984767 690:0.984767 691:0.984767 692:0.984767 693:0.984767 694:0.984767 695:0.984767 696:0.984767 697:0.984767 698:0.984767 699:0.984767 700:0.984767 701:0.984767 702:0.984767 703:0.984767 704:0.984767 705:0.984767 706:0.984767 707:0.984767 708:0.984767 709:0.984767 710:0.984767 711:0.984767 712:0.984767 713:0.984767 714:0.984767 715:0.984767 716:0.984767 717:0.984767 718:0.984767 719:0.984767 720:0.984767 721:0.984767 722:0.984767 723:0.984767 724:0.984767 725:0.984767 726:0.984767 727:0.984767 728:0.984767 729:0.984767 730:0.984767 731:0.984767 732:0.984767 733:0.984767 734:0.984767 735:0.984767 736:0.984767 737:0.984767 738:0.984767 739:0.984767 740:0.984767 741:0.984767 742:0.984767 743:0.984767 744:0.984767 745:0.984767 746:0.984767 747:0.984767 748:0.984767 749:0.984767 750:0.984767 751:0.984767 752:0.984767 753:0.984767 754:0.984767 755:0.984767 756:0.984767 757:0.984767 758:0.984767 759:0.984767 760:0.984767 761:0.984767 762:0.984767 763:0.984767 764:0.984767 765:0.984767 766:0.984767 767:0.984767 768:0.984767 769:0.984767 770:0.984767 771:0.984767 772:0.984767 773:0.984767 774:0.984767 775:0.984767 776:0.984767 777:0.984767 778:0.984767 779:0.984767 780:0.984767 781:0.984767 782:0.984767 783:0.984767 784:0.984767 785:0.984767 786:0.984767 787:0.984767 788:0.984767 789:0.984767 790:0.984767 791:0.984767 792:0.984767 793:0.984767 794:0.984767 795:0.984767 796:0.984767 797:0.984767 798:0.984767 799:0.984767
grad_clip_norm = 100.000000

augment_symmetries = 1 0 0
augment_rotation = vertical
augment_noise = 0.001000
augment_occlusion = none
augment_occlusion_ratio = 0.200000
augment_occlusion_num = 1
augment_scale_anisotropic = 1
augment_scale_min = 0.800000
augment_scale_max = 1.200000
augment_color = 0.800000

weight_decay = 0.001000
segloss_balance = none
class_w =
deform_fitting_mode = point2point
deform_fitting_power = 1.000000
deform_lr_factor = 0.100000
repulse_extent = 1.200000
batch_num = 4
val_batch_num = 4
max_epoch = 800
epoch_steps = 500
validation_size = 200
checkpoint_gap = 50


utils/config.py
#
#
#      0=================================0
#      |    Kernel Point Convolutions    |
#      0=================================0
#
#
# ----------------------------------------------------------------------------------------------------------------------
#
#      Configuration class
#
# ----------------------------------------------------------------------------------------------------------------------
#
#      Hugues THOMAS - 11/06/2018
#


from os.path import join
import numpy as np


# Colors for printing
class bcolors:
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'


class Config:
    """
    Class containing the parameters you want to modify for this dataset
    """

    ##################
    # Input parameters
    ##################

    # Dataset name
    dataset = ''

    # Type of network model
    dataset_task = ''

    # Number of classes in the dataset
    num_classes = 0

    # Dimension of input points
    in_points_dim = 3

    # Dimension of input features
    in_features_dim = 1

    # Radius of the input sphere (ignored for models, only used for point clouds)
    in_radius = 1.0

    # Number of CPU threads for the input pipeline
    input_threads = 8

    ##################
    # Model parameters
    ##################

    # Architecture definition. List of blocks
    architecture = []

    # Decide the mode of equivariance and invariance
    equivar_mode = ''
    invar_mode = ''

    # Dimension of the first feature maps
    first_features_dim = 64

    # Batch normalization parameters
    use_batch_norm = True
    batch_norm_momentum = 0.99

    # For segmentation models : ratio between the segmented area and the input area
    segmentation_ratio = 1.0

    ###################
    # KPConv parameters
    ###################

    # Number of kernel points
    num_kernel_points = 15

    # Size of the first subsampling grid in meter
    first_subsampling_dl = 0.02

    # Radius of convolution in "number grid cell". (2.5 is the standard value)
    conv_radius = 2.5

    # Radius of deformable convolution in "number grid cell". Larger so that deformed kernel can spread out
    deform_radius = 5.0

    # Kernel point influence radius
    KP_extent = 1.0

    # Influence function when d < KP_extent. ('constant', 'linear', 'gaussian') When d > KP_extent, always zero
    KP_influence = 'linear'

    # Aggregation function of KPConv in ('closest', 'sum')
    # Decide if you sum all kernel point influences, or if you only take the influence of the closest KP
    aggregation_mode = 'sum'

    # Fixed points in the kernel : 'none', 'center' or 'verticals'
    fixed_kernel_points = 'center'

    # Use modulateion in deformable convolutions
    modulated = False

    # For SLAM datasets like SemanticKitti number of frames used (minimum one)
    n_frames = 1

    # For SLAM datasets like SemanticKitti max number of point in input cloud + validation
    max_in_points = 0
    val_radius = 51.0
    max_val_points = 50000

    #####################
    # Training parameters
    #####################

    # Network optimizer parameters (learning rate and momentum)
    learning_rate = 1e-3
    momentum = 0.9

    # Learning rate decays. Dictionary of all decay values with their epoch {epoch: decay}.
    lr_decays = {200: 0.2, 300: 0.2}

    # Gradient clipping value (negative means no clipping)
    grad_clip_norm = 100.0

    # Augmentation parameters
    augment_scale_anisotropic = True
    augment_scale_min = 0.9
    augment_scale_max = 1.1
    augment_symmetries = [False, False, False]
    augment_rotation = 'vertical'
    augment_noise = 0.005
    augment_color = 0.7

    # Augment with occlusions (not implemented yet)
    augment_occlusion = 'none'
    augment_occlusion_ratio = 0.2
    augment_occlusion_num = 1

    # Regularization loss importance
    weight_decay = 1e-3

    # The way we balance segmentation loss DEPRECATED
    segloss_balance = 'none'

    # Choose weights for class (used in segmentation loss). Empty list for no weights
    class_w = []

    # Deformable offset loss
    # 'point2point' fitting geometry by penalizing distance from deform point to input points
    # 'point2plane' fitting geometry by penalizing distance from deform point to input point triplet (not implemented)
    deform_fitting_mode = 'point2point'
    deform_fitting_power = 1.0              # Multiplier for the fitting/repulsive loss
    deform_lr_factor = 0.1                  # Multiplier for learning rate applied to the deformations
    repulse_extent = 1.0                    # Distance of repulsion for deformed kernel points

    # Number of batch
    batch_num = 10
    val_batch_num = 10

    # Maximal number of epochs
    max_epoch = 1000

    # Number of steps per epochs
    epoch_steps = 1000

    # Number of validation examples per epoch
    validation_size = 100

    # Number of epoch between each checkpoint
    checkpoint_gap = 50

    # Do we nee to save convergence
    saving = True
    saving_path = None

    #####################
    ## VLAD parameters ##
    #####################

    # seg_net log
    KPlog = ''
    # num of KP features used
    num_feat = 5
    # optimiser
    optimiser = 'Adam'
    # loss function
    loss = 'lazy_quadruplet'
    # neg sample number
    num_neg_samples = 8


    def __init__(self):
        """
        Class Initialyser
        """

        # Number of layers
        self.num_layers = len([block for block in self.architecture if 'pool' in block or 'strided' in block]) + 1

        ###################
        # Deform layer list
        ###################
        #
        # List of boolean indicating which layer has a deformable convolution
        #

        layer_blocks = []
        self.deform_layers = []
        arch = self.architecture
        for block_i, block in enumerate(arch):

            # Get all blocks of the layer
            if not ('pool' in block or 'strided' in block or 'global' in block or 'upsample' in block):
                layer_blocks += [block]
                continue

            # Convolution neighbors indices
            # *****************************

            deform_layer = False
            if layer_blocks:
                if np.any(['deformable' in blck for blck in layer_blocks]):
                    deform_layer = True

            if 'pool' in block or 'strided' in block:
                if 'deformable' in block:
                    deform_layer = True

            self.deform_layers += [deform_layer]
            layer_blocks = []

            # Stop when meeting a global pooling or upsampling
            if 'global' in block or 'upsample' in block:
                break

    def load(self, path):

        filename = join(path, 'parameters.txt')
        with open(filename, 'r') as f:
            lines = f.readlines()

        # Class variable dictionary
        for line in lines:
            line_info = line.split()
            if len(line_info) > 2 and line_info[0] != '#':

                if line_info[2] == 'None':
                    setattr(self, line_info[0], None)

                elif line_info[0] == 'lr_decay_epochs':
                    self.lr_decays = {int(b.split(':')[0]): float(b.split(':')[1]) for b in line_info[2:]}

                elif line_info[0] == 'architecture':
                    self.architecture = [b for b in line_info[2:]]

                elif line_info[0] == 'augment_symmetries':
                    self.augment_symmetries = [bool(int(b)) for b in line_info[2:]]

                elif line_info[0] == 'num_classes':
                    if len(line_info) > 3:
                        self.num_classes = [int(c) for c in line_info[2:]]
                    else:
                        self.num_classes = int(line_info[2])

                elif line_info[0] == 'class_w':
                    self.class_w = [float(w) for w in line_info[2:]]

                elif hasattr(self, line_info[0]):
                    attr_type = type(getattr(self, line_info[0]))
                    if attr_type == bool:
                        setattr(self, line_info[0], attr_type(int(line_info[2])))
                    else:
                        setattr(self, line_info[0], attr_type(line_info[2]))

        self.saving = True
        self.saving_path = path
        self.__init__()

    def save(self):

        with open(join(self.saving_path, 'parameters.txt'), "w") as text_file:

            text_file.write('# -----------------------------------#\n')
            text_file.write('# Parameters of the training session #\n')
            text_file.write('# -----------------------------------#\n\n')

            # Input parameters
            text_file.write('# Input parameters\n')
            text_file.write('# ****************\n\n')
            text_file.write('dataset = {:s}\n'.format(self.dataset))
            text_file.write('dataset_task = {:s}\n'.format(self.dataset_task))
            if type(self.num_classes) is list:
                text_file.write('num_classes =')
                for n in self.num_classes:
                    text_file.write(' {:d}'.format(n))
                text_file.write('\n')
            else:
                text_file.write('num_classes = {:d}\n'.format(self.num_classes))
            text_file.write('in_points_dim = {:d}\n'.format(self.in_points_dim))
            text_file.write('in_features_dim = {:d}\n'.format(self.in_features_dim))
            text_file.write('in_radius = {:.6f}\n'.format(self.in_radius))
            text_file.write('input_threads = {:d}\n\n'.format(self.input_threads))

            # Model parameters
            text_file.write('# Model parameters\n')
            text_file.write('# ****************\n\n')

            text_file.write('architecture =')
            for a in self.architecture:
                text_file.write(' {:s}'.format(a))
            text_file.write('\n')
            text_file.write('equivar_mode = {:s}\n'.format(self.equivar_mode))
            text_file.write('invar_mode = {:s}\n'.format(self.invar_mode))
            text_file.write('num_layers = {:d}\n'.format(self.num_layers))
            text_file.write('first_features_dim = {:d}\n'.format(self.first_features_dim))
            text_file.write('use_batch_norm = {:d}\n'.format(int(self.use_batch_norm)))
            text_file.write('batch_norm_momentum = {:.6f}\n\n'.format(self.batch_norm_momentum))
            text_file.write('segmentation_ratio = {:.6f}\n\n'.format(self.segmentation_ratio))

            # KPConv parameters
            text_file.write('# KPConv parameters\n')
            text_file.write('# *****************\n\n')

            text_file.write('first_subsampling_dl = {:.6f}\n'.format(self.first_subsampling_dl))
            text_file.write('num_kernel_points = {:d}\n'.format(self.num_kernel_points))
            text_file.write('conv_radius = {:.6f}\n'.format(self.conv_radius))
            text_file.write('deform_radius = {:.6f}\n'.format(self.deform_radius))
            text_file.write('fixed_kernel_points = {:s}\n'.format(self.fixed_kernel_points))
            text_file.write('KP_extent = {:.6f}\n'.format(self.KP_extent))
            text_file.write('KP_influence = {:s}\n'.format(self.KP_influence))
            text_file.write('aggregation_mode = {:s}\n'.format(self.aggregation_mode))
            text_file.write('modulated = {:d}\n'.format(int(self.modulated)))
            text_file.write('n_frames = {:d}\n'.format(self.n_frames))
            text_file.write('max_in_points = {:d}\n\n'.format(self.max_in_points))
            text_file.write('max_val_points = {:d}\n\n'.format(self.max_val_points))
            text_file.write('val_radius = {:.6f}\n\n'.format(self.val_radius))

            # Training parameters
            text_file.write('# Training parameters\n')
            text_file.write('# *******************\n\n')

            text_file.write('learning_rate = {:f}\n'.format(self.learning_rate))
            text_file.write('momentum = {:f}\n'.format(self.momentum))
            text_file.write('lr_decay_epochs =')
            for e, d in self.lr_decays.items():
                text_file.write(' {:d}:{:f}'.format(e, d))
            text_file.write('\n')
            text_file.write('grad_clip_norm = {:f}\n\n'.format(self.grad_clip_norm))


            text_file.write('augment_symmetries =')
            for a in self.augment_symmetries:
                text_file.write(' {:d}'.format(int(a)))
            text_file.write('\n')
            text_file.write('augment_rotation = {:s}\n'.format(self.augment_rotation))
            text_file.write('augment_noise = {:f}\n'.format(self.augment_noise))
            text_file.write('augment_occlusion = {:s}\n'.format(self.augment_occlusion))
            text_file.write('augment_occlusion_ratio = {:.6f}\n'.format(self.augment_occlusion_ratio))
            text_file.write('augment_occlusion_num = {:d}\n'.format(self.augment_occlusion_num))
            text_file.write('augment_scale_anisotropic = {:d}\n'.format(int(self.augment_scale_anisotropic)))
            text_file.write('augment_scale_min = {:.6f}\n'.format(self.augment_scale_min))
            text_file.write('augment_scale_max = {:.6f}\n'.format(self.augment_scale_max))
            text_file.write('augment_color = {:.6f}\n\n'.format(self.augment_color))

            text_file.write('weight_decay = {:f}\n'.format(self.weight_decay))
            text_file.write('segloss_balance = {:s}\n'.format(self.segloss_balance))
            text_file.write('class_w =')
            for a in self.class_w:
                text_file.write(' {:.6f}'.format(a))
            text_file.write('\n')
            text_file.write('deform_fitting_mode = {:s}\n'.format(self.deform_fitting_mode))
            text_file.write('deform_fitting_power = {:.6f}\n'.format(self.deform_fitting_power))
            text_file.write('deform_lr_factor = {:.6f}\n'.format(self.deform_lr_factor))
            text_file.write('repulse_extent = {:.6f}\n'.format(self.repulse_extent))
            text_file.write('batch_num = {:d}\n'.format(self.batch_num))
            text_file.write('val_batch_num = {:d}\n'.format(self.val_batch_num))
            text_file.write('max_epoch = {:d}\n'.format(self.max_epoch))
            if self.epoch_steps is None:
                text_file.write('epoch_steps = None\n')
            else:
                text_file.write('epoch_steps = {:d}\n'.format(self.epoch_steps))
            text_file.write('validation_size = {:d}\n'.format(self.validation_size))
            text_file.write('checkpoint_gap = {:d}\n'.format(self.checkpoint_gap))

            # VLAD parameters
            text_file.write('\n')
            text_file.write('KPlog = {:s}\n'.format(self.KPlog))
            text_file.write('num_feat = {:d}\n'.format(self.num_feat))
            text_file.write('optimiser = {:s}\n'.format(self.optimiser))
            text_file.write('loss_function = {:s}\n'.format(self.loss))
            text_file.write('num_neg_samples = {:d}\n'.format(self.num_neg_samples))

    def print_current(self):
        print('\n# -----------------------------------#')
        print('# Parameters of the training session #')
        print('# -----------------------------------#')

        # Input parameters
        print('# Input parameters')
        print('# ****************')
        print('dataset = {:s}'.format(self.dataset))
        print('dataset_task = {:s}'.format(self.dataset_task))
        if type(self.num_classes) is list:
            print('num_classes =', end=' ')
            for n in self.num_classes:
                print('{:d}'.format(n), end=' ')
            print('')
        else:
            print('num_classes = {:d}'.format(self.num_classes))
        print('in_points_dim = {:d}'.format(self.in_points_dim))
        print('in_features_dim = {:d}'.format(self.in_features_dim))
        print('in_radius = {:.6f}'.format(self.in_radius))
        print('input_threads = {:d}'.format(self.input_threads))

        # Model parameters
        print('# Model parameters')
        print('# ****************')
        print('architecture =', end=' ')
        for a in self.architecture:
            if a in ['resnetb_strided', 'resnetb_deformable_strided', 'nearest_upsample']:
                print('\n',' '*13, '{:s}'.format(a), end=' ')
            else:
                print('{:s}'.format(a), end=' ')
        print('')
        print('num_layers = {:d}'.format(self.num_layers))
        print('equivar_mode = {:s}'.format(self.equivar_mode))
        print('invar_mode = {:s}'.format(self.invar_mode))
        print('first_features_dim = {:d}'.format(self.first_features_dim))
        print('use_batch_norm = {:d}'.format(int(self.use_batch_norm)))
        print('batch_norm_momentum = {:.6f}'.format(self.batch_norm_momentum))
        print('segmentation_ratio = {:.6f}'.format(self.segmentation_ratio))

        # KPConv parameters
        print('# KPConv parameters')
        print('# *****************')
        print('num_kernel_points = {:d}'.format(self.num_kernel_points))
        print('first_subsampling_dl = {:.6f}'.format(self.first_subsampling_dl))
        print('conv_radius = {:.6f}'.format(self.conv_radius))
        print('deform_radius = {:.6f}'.format(self.deform_radius))
        print('KP_extent = {:.6f}'.format(self.KP_extent))
        print('KP_influence = {:s}'.format(self.KP_influence))
        print('aggregation_mode = {:s}'.format(self.aggregation_mode))
        print('fixed_kernel_points = {:s}'.format(self.fixed_kernel_points))
        print('modulated = {:d}'.format(int(self.modulated)))
        print('n_frames = {:d}'.format(self.n_frames))
        print('max_in_points = {:d}'.format(self.max_in_points))
        print('max_val_points = {:d}'.format(self.max_val_points))
        print('val_radius = {:.6f}'.format(self.val_radius))

        # Training parameters
        print('# Training parameters')
        print('# *******************')
        print('learning_rate = {:f}'.format(self.learning_rate))
        print('momentum = {:f}'.format(self.momentum))
        print('lr_decay_epochs = [hide for better visual]')
        # for e, d in self.lr_decays.items():
        #     print(' {:d}:{:f}'.format(e, d))
        # print('')
        print('grad_clip_norm = {:f}'.format(self.grad_clip_norm))
        print('augment_scale_anisotropic = {:d}'.format(int(self.augment_scale_anisotropic)))
        print('augment_scale_min = {:.6f}'.format(self.augment_scale_min))
        print('augment_scale_max = {:.6f}'.format(self.augment_scale_max))
        print('augment_symmetries =', end=' ')
        for a in self.augment_symmetries:
            print('{:d}'.format(int(a)), end=' ')
        print('')
        print('augment_rotation = {:s}'.format(self.augment_rotation))
        print('augment_noise = {:f}'.format(self.augment_noise))
        print('augment_color = {:.6f}'.format(self.augment_color))
        print('augment_occlusion = {:s}'.format(self.augment_occlusion))
        print('augment_occlusion_ratio = {:.6f}'.format(self.augment_occlusion_ratio))
        print('augment_occlusion_num = {:d}'.format(self.augment_occlusion_num))
        print('weight_decay = {:f}'.format(self.weight_decay))
        print('segloss_balance = {:s}'.format(self.segloss_balance))
        print('class_w =', end=' ')
        for a in self.class_w:
            print('{:.6f}'.format(a), end=' ')
        print('')
        print('deform_fitting_mode = {:s}'.format(self.deform_fitting_mode))
        print('deform_fitting_power = {:.6f}'.format(self.deform_fitting_power))
        print('deform_lr_factor = {:.6f}'.format(self.deform_lr_factor))
        print('repulse_extent = {:.6f}'.format(self.repulse_extent))
        print('batch_num = {:d}'.format(self.batch_num))
        print('val_batch_num = {:d}'.format(self.val_batch_num))
        print('max_epoch = {:d}'.format(self.max_epoch))
        if self.epoch_steps is None:
            print('epoch_steps = None')
        else:
            print('epoch_steps = {:d}'.format(self.epoch_steps))
        print('validation_size = {:d}'.format(self.validation_size))
        print('checkpoint_gap = {:d}'.format(self.checkpoint_gap))
        print('saving = {:d}'.format(self.saving))
        print('saving_path = {:s}'.format(self.saving_path))

        # VLAD parameters
        print('')
        print('KPlog = {:s}'.format(self.KPlog))
        print('num_feat = {:d}'.format(self.num_feat))
        print('optimiser = {:s}'.format(self.optimiser))
        print('loss_function = {:s}\n'.format(self.loss))
        print('num_neg_samples = {:d}\n'.format(self.num_neg_samples))
        print('# -----------------------------------#\n')

